File: samlang-parser/src/main/kotlin/samlang/parser/PatternBuilder.kt
package samlang.parser

import samlang.ast.lang.Pattern
import samlang.parser.generated.PLBaseVisitor
import samlang.parser.generated.PLParser

internal object PatternBuilder : PLBaseVisitor<Pattern?>() {

    override fun visitTuplePattern(ctx: PLParser.TuplePatternContext): Pattern = Pattern.TuplePattern(
        range = ctx.range,
        destructedNames = ctx.varOrWildCard().map { c ->
            val range = c.range
            c.LowerId()?.symbol?.text to range
        }
    )

    private object FieldNameBuilder : PLBaseVisitor<Pattern.ObjectPattern.DestructedName?>() {

        override fun visitRawVar(ctx: PLParser.RawVarContext): Pattern.ObjectPattern.DestructedName =
            ctx.LowerId().symbol.let {
                Pattern.ObjectPattern.DestructedName(
                    fieldName = it.text,
                    fieldOrder = -1,
                    alias = null,
                    range = it.range
                )
            }

        override fun visitRenamedVar(ctx: PLParser.RenamedVarContext): Pattern.ObjectPattern.DestructedName {
            val idList = ctx.LowerId()
            return Pattern.ObjectPattern.DestructedName(
                fieldName = idList[0].symbol.text,
                fieldOrder = -1,
                alias = idList[1].symbol.text,
                range = ctx.range
            )
        }
    }

    override fun visitObjectPattern(ctx: PLParser.ObjectPatternContext): Pattern? {
        val destructedNames = ctx.varOrRenamedVar().map { it.accept(FieldNameBuilder) ?: return null }
        return Pattern.ObjectPattern(range = ctx.range, destructedNames = destructedNames)
    }

    override fun visitVariablePattern(ctx: PLParser.VariablePatternContext): Pattern =
        Pattern.VariablePattern(range = ctx.range, name = ctx.text)

    override fun visitWildcardPattern(ctx: PLParser.WildcardPatternContext): Pattern =
        Pattern.WildCardPattern(range = ctx.range)
}

File: samlang-parser-generated-ts/src/pattern-builder.ts
import { PLVisitor } from './generated/PLVisitor';
import { AbstractParseTreeVisitor } from 'antlr4ts/tree/AbstractParseTreeVisitor';
import {
  TuplePatternContext,
  ObjectPatternContext,
  RawVarContext,
  RenamedVarContext,
  VariablePatternContext,
  WildcardPatternContext,
} from './generated/PLParser';
import {
  TsPattern,
  TsTuplePattern,
  TsObjectPattern,
  TsObjectDestructedName,
  TsVariablePattern,
  TsWildCardPattern,
} from './ast';
import { tokenRange, contextRange, throwParserError } from './parser-util';

class FieldNameBuilder extends AbstractParseTreeVisitor<TsObjectDestructedName>
  implements PLVisitor<TsObjectDestructedName> {
  defaultResult = (): TsObjectDestructedName => null!;

  visitRawVar = (ctx: RawVarContext): TsObjectDestructedName => {
    const symbol = ctx.LowerId().symbol;
    const fieldName = symbol.text ?? throwParserError('Missing field name in object pattern.');
    return { fieldName, fieldOrder: -1, range: tokenRange(symbol) };
  };

  visitRenamedVar = (ctx: RenamedVarContext): TsObjectDestructedName => {
    const idList = ctx.LowerId();
    const fieldName =
      idList[0].symbol.text ?? throwParserError('Missing field name in object pattern.');
    const alias = idList[1].symbol.text ?? throwParserError('Missing alias in object pattern.');
    return {
      fieldName,
      fieldOrder: -1,
      alias,
      range: contextRange(ctx),
    };
  };
}

const fieldNameBuilder = new FieldNameBuilder();

class PatternBuilder extends AbstractParseTreeVisitor<TsPattern> implements PLVisitor<TsPattern> {
  defaultResult = (): TsPattern => null!;

  visitTuplePattern = (ctx: TuplePatternContext): TsPattern =>
    new TsTuplePattern(
      contextRange(ctx),
      ctx.varOrWildCard().map((c) => ({ name: c.LowerId()?.symbol?.text, range: contextRange(c) }))
    );

  visitObjectPattern = (ctx: ObjectPatternContext): TsPattern => {
    const destructedNames = ctx.varOrRenamedVar().map((it) => it.accept(fieldNameBuilder));
    return new TsObjectPattern(contextRange(ctx), destructedNames);
  };

  visitVariablePattern = (ctx: VariablePatternContext): TsPattern =>
    new TsVariablePattern(contextRange(ctx), ctx.text);

  visitWildcardPattern = (ctx: WildcardPatternContext): TsPattern =>
    new TsWildCardPattern(contextRange(ctx));
}

const patternBuilder = new PatternBuilder();
export default patternBuilder;
